### DISCUSSION TITLE
How to trigger Ipc Server Triggers

### DISCUSSION THREAD

**Discussion by roblouie** [devcompl/TTController#97] [Configuration] [answered]
I want an easy way to change the color of the fan lights, sort of like having multiple profiles I can switch between. From reading the docs, it looks like the easiest way to do this is Ipc Server Triggers? That way i can just post a message to switch at any time...right?

So, I enabled that, and then just for testing fired up a websocket tester, and it says I connected, but when I send in `{ "Enabled": true }`, nothing happens.

Can you help me understand what's going on?

Also thank you so much for making this. Its so much better than the thermaltake software, despite having no UI at all. Which, really is amazing to think about, but it's 100% true.

Here's me connected to it and posting the message:
![image](https://user-images.githubusercontent.com/5448110/101999191-2f3ea280-3ca0-11eb-9b25-ff6e2c6346a0.png)

I also tried it like this:
![image](https://user-images.githubusercontent.com/5448110/101999156-d8d16400-3c9f-11eb-93e2-6f5fc117188b.png)

Here's my config:
```
{
  "Profiles": [
    {
    "Name": "Radiator",
    "Ports": [
      [9802, 8800, 1],
      [9802, 8800, 2],
      [9802, 8800, 3]
    ],
    "SpeedControllers": [
      {
        "Type": "PwmSpeedController", 
        "Config": {
          "CurvePoints": [
            [30, 30],
            [45, 50],
            [55, 60],
            [65, 75],
            [75, 100]
          ],
          "Sensors": ["/amdcpu/0/temperature/2"],
          "Trigger": {
            "Type": "AlwaysTrigger"
          }
        }
      }
    ],
    "Effects": [
      {
        "Type": "StaticColorEffect",
        "Config": {
          "Color": {
            "Full": [50, 50, 50]
          },
          "Trigger": {
            "Type": "AlwaysTrigger"
          }
        }
      },
      {
        "Type": "StaticColorEffect",
        "Config": {
          "Color": {
            "Full": [255, 0, 20]
          },
          "Trigger": {
            "Type": "IpcTrigger",
            "Config": {
              "IpcName": "cyberpunk",
              "EnabledByDefault": false
            }
          }
        }
      }
    ]
  },
  {
    "Name": "Case",
    "Ports": [
      [9802, 8801, 3],
      [9802, 8800, 4],
      [9802, 8800, 5]
    ],
    "SpeedControllers": [
      {
        "Type": "StaticSpeedController", 
        "Config": {
          "Speed": 20,
          "Trigger": {
            "Type": "AlwaysTrigger"
          }
        }
      }
    ],
    "Effects": [
      {
        "Type": "StaticColorEffect",
        "Config": {
          "Color": {
            "Full": [50, 50, 50]
          },
          "Trigger": {
            "Type": "AlwaysTrigger"
          }
        }
      }
    ]
  }
],
  "ComputerStateProfiles": [],
  "PortConfigs": [{
    "Ports": [
      [9802, 8800, 1],
      [9802, 8800, 2],
      [9802, 8800, 3],
      [9802, 8800, 4],
      [9802, 8800, 5],
      [9802, 8801, 3]
    ],
    "Config": {
      "DeviceType": "RiingQuad",
      "IgnoreColorCache": true
    }
  }],
  "SensorConfigs": [],
  "IpcServer": {
    "Type": "WebSocketIpcServer",
    "Config": {
      "Address": "127.0.0.1"
    }
  },
  "IpcServerEnabled": true
}
```

**Comment by roblouie** [comment]
As a note, I did also try removing the "AlwaysTrigger" from my default color palette, just in case that was overriding what i sent in. Still didn't work. Here's the relevant part of that config:

```
"Effects": [
      {
        "Type": "StaticColorEffect",
        "Config": {
          "Color": {
            "Full": [50, 50, 50]
          },
         "Trigger": {
            "Type": "IpcTrigger",
            "Config": {
              "IpcName": "default",
              "EnabledByDefault": true
            }
          }
        }
      },
      {
        "Type": "StaticColorEffect",
        "Config": {
          "Color": {
            "Full": [255, 0, 20]
          },
          "Trigger": {
            "Type": "IpcTrigger",
            "Config": {
              "IpcName": "cyberpunk",
              "EnabledByDefault": false
            }
          }
        }
      }
    ]
  },
```

**Comment by MoshiMoshi0** [comment, marked as answer]
> I want an easy way to change the color of the fan lights, sort of like having multiple profiles I can switch between. From reading the docs, it looks like the easiest way to do this is Ipc Server Triggers? That way i can just post a message to switch at any time...right?

Yes, the only way to "control" the service is with ipc plugins. Other triggers could also be used but they are more event based, like process running or time of day. 
Ipc plugins were not yet user tested and I think you are the first to use them, so let me know if you have any suggestions. Just by testing now I found few small issues.

Back to the issue.
The plugin triggers are evaluated from first to last. First one to return true marks that plugin as "enabled".
So in your OP config the `"StaticColorEffect"` with `"AlwaysTrigger"` was always being used, without even checking the next effect.

Fixed should look like this:
```json
    "Effects": [
      {
        "Type": "StaticColorEffect",
        "Config": {
          "Color": {
            "Full": [255, 0, 20]
          },
          "Trigger": {
            "Type": "IpcTrigger",
            "Config": {
              "IpcName": "cyberpunk",
              "EnabledByDefault": false
            }
          }
        }
      },
      {
        "Type": "StaticColorEffect",
        "Config": {
          "Color": {
            "Full": [50, 50, 50]
          },
          "Trigger": {
            "Type": "AlwaysTrigger"
          }
        }
      }
    ]
```

Tested and it works correctly for me when sending `{ "Enabled": false }` or `{ "Enabled": true }`.

Also it seems that I forgot to mention on the docs how is `"IpcName"` used but you figured it out. With websocket server you connect to `ws://<address>:<port>/<ipc name>`, the `ws://127.0.0.1:8888/service` client you connected to does not output any data unless [IpcClientTimerInterval](https://moshimoshi0.github.io/TTController/config/service#ipcclienttimerinterval) is set. It just dumps all port data (color + speed) for now, but I planned for it to be used to also list available ipc plugins and other stuff.
So you should connect to `ws://127.0.0.1:8888/cyberpunk`.

Best way to test is to run the service in console mode, it will output additional debug stuff.

**Comment by roblouie** [comment]
Thank you so much, that got it.
